encryption-at-rest: Update (#6152)
pingcap/docs: encryption-at-rest.md

TiKV supports encryption at rest. The feature allows TiKV to transparently encrypt data files using [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) in [CTR](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation)  mode. To enable encryption at rest, an encryption key must be provided by user and this key is called master key. The master key can be provided via AWS KMS (recommended), or specifying a key stored as plaintext in a file. TiKV automatically rotates data keys that it used to encrypt actual data files. Manually rotating the master key can be done occasionally. Note that encryption at rest only encrypts data at rest (namely, on disk) and not while data is transferred over network. It is advised to use TLS together with encryption at rest.

> TiFlash also supports encryption at rest and while it only stores some metadata PD also supports data at rest. Encryption at rest can be enabled per component.

Reviewer: Please note that PD's encryption-at-rest support is experimental.

Author: I've added a warning for that

TiFlash also supports encryption at rest and while it only stores some metadata PD also supports data at rest. Encryption at rest can be enabled per component.

Using AWS KMS is also possible for on-premise deployments.

BR supports S3 server-side encryption (SSE) when backing up to S3. A customer owned AWS KMS key can also be used together with S3 server-side encryption.

## Warnings

The current version of TiKV encryption has the following drawbacks. Be aware of these drawbacks before you get started:

* When a TiDB cluster is deployed, the majority of user data is stored in TiKV and TiFlash nodes, and that data will be encrypted when encryption is enabled. However, a small amount of user data is stored in PD nodes as metadata (for example, secondary index keys used as TiKV region boundaries). PD also supports encryption at rest.

> * TiFlash supports encryption at rest, for details, refer to [Encryption at Rest for TiFlash](#encryption-at-rest-for-tiflash).

Reviewer: I think we can refactor this a bit. We can remove these two warning items, but add a dedicated paragraph somewhere noting encryption should be enable per component. We can merge the warning of PD encryption being experimental here.

Author: Would it be ok to merge this PR and then let you take care of refactoring it? Or could you add a suggestion?

aliases: ['/docs/dev/encryption at rest/']
---

# Encryption at Rest

> **Note:**
>
> When deploying on AWS it is recommened to consider using [EBS encryption](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSEncryption.html) instead unless you are using non-EBS storage like local NVMe storage.

Encryption at rest means that data is encrypted when it is stored. For databases, this feature is also referred to as TDE (transparent data encryption). This is opposed to encryption in flight (TLS) or encryption in use (rarely used). Different things could be doing encryption at rest (SSD drive, file system, cloud vendor, etc), but by having TiKV do the encryption before storage this helps ensure that attackers must authenticate with the database to gain access to data. For example, when an attacker gains access to the physical machine, data cannot be accessed by copying files on disk.

## Per component encryption support

> When a TiDB cluster is deployed, the majority of user data is stored in TiKV and TiFlash nodes. Some metadata is stored in PD nodes (for example, secondary index keys used as TiKV region boundaries). To get the full benefits of this feature encryption needs to be enabled for all components. Backups, log files and data transmitted over the network should also be considered when implementing this.

Reviewer: ```suggestion
When a TiDB cluster is deployed, the majority of user data is stored in TiKV and TiFlash nodes. Some metadata is stored in PD nodes (for example, secondary index keys used as TiKV region boundaries). To get the full benefits of this feature encryption needs to be enabled for all components. Backups and data transmitted over the network (TLS) should also be considered when implementing this.
```
We don't have feature to encrypt the log files, so probably don't mention it.

Author: But there is https://docs.pingcap.com/tidb/stable/log-redaction and the option to change the log level.

Reviewer: got it. specifically mention log redaction and link to the docs? also for TLS.

Author: There is a `Logging` section that mentions log redaction and links to it.

As logging is stored on disk (at rest) and TLS is not I decided to not go into much detail on TLS here.

### TiKV

> TiKV supports encryption at rest, this allows TiKV to transparently encrypt data files using [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) in [CTR](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation) mode. To enable encryption at rest, an encryption key must be provided by user and this key is called master key. The master key can be provided via AWS KMS (recommended), or specifying a key stored as plaintext in a file. TiKV automatically rotates data keys that it used to encrypt actual data files. Manually rotating the master key can be done occasionally. Note that encryption at rest only encrypts data at rest (namely, on disk) and not while data is transferred over network. It is advised to use TLS together with encryption at rest

Author: Updated. PTAL

> It is possible to use AWS KMS for on-premise deployments, but this is not required.

Author: Updated. PTAL

TiKV currently does not exclude encryption keys and user data from core dumps. It is advised to disable core dumps for the TiKV process when using encryption at rest. This is not currently handled by TiKV itself.

TiKV tracks encrypted data files using the absolute path of the files. As a result, once encryption is turned on for a TiKV node, the user should not change data file paths configuration such as `storage.data-dir`, `raftstore.raftdb-path`, `rocksdb.wal-dir` and `raftdb.wal-dir`.

### TiFlash

TiFlash supports encryption at rest. Data keys are generated by TiFlash. All files (including data files, schema files, and temporary files) written into TiFlash (including TiFlash Proxy) are encrypted using the current data key. The encryption algorithms, the encryption configuration (in the `tiflash-learner.toml` file) supported by TiFlash, and the meanings of monitoring metrics are consistent with those of TiKV.

If you have deployed TiFlash with Grafana, you can check the **TiFlash-Proxy-Details** -> **Encryption** panel.

### PD

Using encryption-at-rest for PD is an experimental feature. This is configured in the same way as TiKV.

### Backups with BR

> BR supports S3 server-side encryption (SSE) when backing up to S3. A customer owned AWS KMS key can also be used together with S3 server-side encryption. See [BR S3 server-side encryption](/encryption-at-rest.md#br-s3-server-side-encryption) for details.

Reviewer 2: ```suggestion
BR supports S3 server-side encryption (SSE) when backing up data to S3. A customer-owned AWS KMS key can also be used together with S3 server-side encryption. See [BR S3 server-side encryption](/encryption-at-rest.md#br-s3-server-side-encryption) for details.
```

* TiFlash supports encryption at rest since v4.0.5. For details, refer to [Encryption at Rest for TiFlash](#encryption-at-rest-for-tiflash-new-in-v405). When deploying TiKV with TiFlash earlier than v4.0.5, data stored in TiFlash is not encrypted.
* TiKV currently does not exclude encryption keys and user data from core dumps. It is advised to disable core dumps for the TiKV process when using encryption at rest. This is not currently handled by TiKV itself.
* TiKV tracks encrypted data files using the absolute path of the files. As a result, once encryption is turned on for a TiKV node, the user should not change data file paths configuration such as `storage.data-dir`, `raftstore.raftdb-path`, `rocksdb.wal-dir` and `raftdb.wal-dir`.

> * TiKV, TiDB, and PD  info logs might contain user data for debugging purposes. The info log and this data in it are not encrypted. It is recommended to enable [log redaction](/log-redaction.md).

Reviewer: Let's keep the last three warning items. They are still valid.

Author: These were not removed, just moved around.

Reviewer: I see.

## Encryption support in different TiDB components

In a TiDB cluster, different components use different encryption methods. This section introduces the encryption supports in different TiDB components such as TiKV, TiFlash, PD, and Backup & Restore (BR).

When a TiDB cluster is deployed, the majority of user data is stored on TiKV and TiFlash nodes. Some metadata is stored on PD nodes (for example, secondary index keys used as TiKV Region boundaries). To get the full benefits of encryption at rest, you need to enable encryption for all components. Backups, log files, and data transmitted over the network should also be considered when you implement encryption.

### TiKV

> TiKV supports encryption at rest, this allows TiKV to transparently encrypt data files using [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) in [CTR](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation) mode. To enable encryption at rest, an encryption key must be provided by user and this key is called master key. The master key that is used to encrypt the data keys is provided by the user. TiKV automatically rotates data keys that it used to encrypt actual data files. Manually rotating the master key can be done occasionally. Note that encryption at rest only encrypts data at rest (namely, on disk) and not while data is transferred over network. It is advised to use TLS together with encryption at rest

Reviewer 2: See the duplicated content.
![image](https://user-images.githubusercontent.com/41534398/132809863-990f4385-0ba8-4f05-ba4b-9cba6532c736.png)

Please make sure the same information is only stated once in this doc if not for emphatic purposes.

Author: Like this? 1a5b3c5e32db6ce6825dc3b97dc3eef697efbcef

Reviewer 2: LGTM

Reviewer 2: ```suggestion
TiKV supports encryption at rest. This feature allows TiKV to transparently encrypt data files using [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) in [CTR](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation) mode. To enable encryption at rest, an encryption key must be provided by the user and this key is called master key. TiKV automatically rotates data keys that it used to encrypt actual data files. Manually rotating the master key can be done occasionally. Note that encryption at rest only encrypts data at rest (namely, on disk) and not while data is transferred over network. It is advised to use TLS together with encryption at rest.
```

> AWS KMS can be used for both Cloud and on-premise deployments, but this is not required. It is also possible to supply the plaintext master key in a file.

Reviewer 2: ```suggestion
Optionally, you can use AWS KMS for both cloud and on-premises deployments. You can also supply the plaintext master key in a file.
```

Regardless of data encryption method, data keys are encrypted using AES256 in GCM mode for additional authentication. This required the master key to be 256 bits (32 bytes), when passing from file instead of KMS.

### Key creation

> Go to the [AWS KMS](https://console.aws.amazon.com/kms) on the AWS console. Make sure the correct region is selected on the top right corner of your console. Make  and click "Create a key". Select "Symmetric" as Key type. After this you can set an alias an description and set tags.

Reviewer 2: ```suggestion

To create a key on AWS, follow these steps:

1. Go to the [AWS KMS](https://console.aws.amazon.com/kms) on the AWS console.
2. Make sure that you have selected the correct region on the top right corner of your console.
3. Click **Create a key** and select **Symmetric** as the key type.
4. Set an alias for the key and set tags.
```

Reviewer 2: @dveeden PTAL

It is also possible to do this with the AWS Cli:

> ```

Reviewer 2: ```suggestion
```shell
```

endpoint = "https://kms.us-west-2.amazonaws.com"
```

The `key-id` specifies the key id for the KMS CMK. The `region` is the AWS region name for the KMS CMK. The `endpoint` is optional and doesn't need to be specified normally, unless you are using a AWS KMS compatible service from a non-AWS vendor or need to use a [VPC endpoint for KMS](https://docs.aws.amazon.com/kms/latest/developerguide/kms-vpc-endpoint.html).

> It is possible to use [multi-Region keys](https://docs.aws.amazon.com/kms/latest/developerguide/multi-region-keys-overview.html). For this you need to setup a primary key in a specific region and add replica keys in the regions you require.

Reviewer 2: ```suggestion
You can also use [multi-Region keys](https://docs.aws.amazon.com/kms/latest/developerguide/multi-region-keys-overview.html) in AWS. For this, you need to set up a primary key in a specific region and add replica keys in the regions you require.
```
